src/main/scala/za/co/absa/abris/avro/sql/AvroDataToCatalyst.scala

```diff
@@ -46,13 +46,25 @@ case class AvroDataToCatalyst(
 
   override def nullable: Boolean = true
 
-  @transient private lazy val schemaManager = SchemaManagerFactory.create(schemaRegistryConf.get)
+  val confluentCompliant = config.confluent
 
-  @transient private lazy val avroSchema = AvroSchemaUtils.parse(jsonFormatSchema)
+  @transient private lazy val schemaManager = SchemaManagerFactory.create(config.schemaRegistryConf.get)
+
+  // Avro schema to be used for reading
+  @transient private lazy val avroSchema = AvroSchemaUtils.parse(config.schemaString)
+
+  // Avro schema that was used for data serialization
+  @transient private lazy val writerSchema = config.writerSchema.map(s => AvroSchemaUtils.parse(s))
+
+  // Reused GenericDatumReaders per writer schema
+  @transient private lazy val vanillaReader: GenericDatumReader[Any] = new GenericDatumReader[Any](writerSchema.getOrElse(avroSchema), avroSchema)
+  @transient private lazy val confluentReaderCache: ConcurrentHashMap[Int, GenericDatumReader[Any]] = new ConcurrentHashMap[Int, GenericDatumReader[Any]]()
```

Comment:
Is ConcurrentHashMap really needed here? I tried to find how exactly the expressions are executed in Spark, but it's difficult to find enough information about this.

- The expression will be executed on executor as part of Spark task. 
- Each task should run only in one thread.
- if there is only one instance of Expression per task it would mean the private variables are accessed only by one thread => no need to worry about thread safety

I looked at the source code and it seems to be that way:
- `Executor` has a thread pool that it uses to run `TaskRunners`
- `TaskRunner` is Runnable that runs one instance of a `Task` at a time
- `Task` is received from the driver and deserialized inside the `TaskRunner`
- `ResultTask` (implementation of `Task`) seems to be deserializing the function that it runs. I think this is a function that will contain this expression.

From this it seems there will be only one instance of AvroDataToCatalyst expression per Executor thread. The instance is created by deserialization in Task. That would mean transient private fields will be accessed only by one thread.

But the Spark code is quite complex, so I might miss something.

Reply:
Yes I have the same understanding that it's only a single thread that is using this an instance of this class. Otherwise we'd probably also run into problems with the cached result and other fields. The implementation in spark-avro does it similarly.

I'll use a `mutable.HashMap` instead.